sql_generators/glean_usage/events_stream.py

```diff
@@ -126,4 +148,72 @@ def generate_per_app(
             parallelism=parallelism,
             id_token=id_token,
             all_base_tables_exist=all_base_tables_exist,
+            custom_render_kwargs=custom_render_kwargs,
         )
+
+
+@cache
+def get_glean_app_repository(v1_name: str) -> dict:
+    """Return the Glean app's repository."""
+    for repository in get_glean_repositories():
+        if repository["name"] == v1_name:
+            return repository
+    raise Exception(f"No Glean repository found for app `{v1_name}`.")
+
+
+@cache
+def get_glean_dependency_v1_name(dependency: str) -> str:
+    """Return the v1 name of the Glean dependency, which may be a library."""
+    for repository in get_glean_repositories():
+        if "library_names" in repository and dependency in repository["library_names"]:
+            return repository["name"]
+    return dependency
+
+
+@cache
+def get_glean_app_event_extras_by_type(
+    v1_name: str, ping: str = "events", cutoff_date: Optional[date] = None
+) -> dict[str, set[str]]:
+    """Return the Glean app's event extra keys for the specified ping, grouped by type."""
+    extras_by_type = defaultdict(set)
+    repository = get_glean_app_repository(v1_name)
+
+    if not cutoff_date:
+        ping_delete_after_days = (
+            (
+                repository.get("moz_pipeline_metadata", {})
+                .get(ping, {})
+                .get("expiration_policy", {})
+                .get("delete_after_days")
+            )
+            or (
+                repository.get("moz_pipeline_metadata_defaults", {})
+                .get("expiration_policy", {})
+                .get("delete_after_days")
+            )
+            or 775
+        )
+        cutoff_date = date.today() - timedelta(days=(ping_delete_after_days - 1))
+
+    metrics = list(get_glean_app_metrics(v1_name).values())
+    for metric in metrics:
+        if metric["type"] == "event":
+            for history in metric["history"]:
+                last_date = datetime.fromisoformat(history["dates"]["last"]).date()
+                if ping in history["send_in_pings"] and last_date >= cutoff_date:
+                    for extra_key, extra_key_info in history["extra_keys"].items():
+                        extra_type = extra_key_info.get("type", "string")
+                        extras_by_type[extra_type].add(extra_key)
+
+    for dependency in repository["dependencies"]:
+        dependency_v1_name = get_glean_dependency_v1_name(dependency)
+        dependency_extras_by_type = get_glean_app_event_extras_by_type(
+            dependency_v1_name, ping, cutoff_date
+        )
+        for (
+            depencency_extra_type,
+            depencency_extras,
+        ) in dependency_extras_by_type.items():
+            extras_by_type[depencency_extra_type].update(depencency_extras)
```

Comment:
type o
```suggestion
        for (
            dependency_extra_type,
            dependency_extras,
        ) in dependency_extras_by_type.items():
            extras_by_type[dependency_extra_type].update(dependency_extras)
```

Reply:
Thanks for catching that!  Can't believe I let that slip through. 😬